### ISSUE TITLE
General conversation

### ISSUE THREAD

**Issue by qwertie** [loycnet/ecsharp#1]
A thread for chatting.

**Comment by qwertie** [comment]
So I'd like to use the word "we" on the home page so as to sound bigger. I'm just going to imagine that the plural is you, Jonathan. Sound good?

**Comment by jonathanvdc** [comment]
Yep. I'm absolutely fine with fine with that. Nothing wrong with sounding bigger.

**Comment by qwertie** [comment]
So, did you write the Flame backend for C++ before you discovered Loyc? I ask because I've been dying to get a conversion from Loyc trees to C++ (enabling surface-syntax conversion from C# to C++, e.g. so that LLLPG could begin supporting C++ as a compilation target), and am curious why you didn't write one.

**Comment by jonathanvdc** [comment]
I'm not entirely sure if I had already discovered Loyc when I wrote Flame's C++ back-end, but I certainly hadn't used Loyc at that point. If I had to do it all over again, then I'd probably build a separate C++ printer for Loyc trees and then use that instead of generating the C++ code directly. Part of why the C++ back-end is such a mess is because it was a tool I needed. I didn't write it to be an additional selling point for Flame. My reasoning for listing it on the Flame README is that it "proves" that Flame is not tied to any specific back-end or platform.

Here's some background on the C++ back-end. My motivation for writing the C++ back-end was that I had to write C++ code, but what I wanted to write was C# code (or D# code, but that's more or less a subset of C#). So I figured that I could just write a C++ back-end for Flame to satisfy all parties involved. And if you look at the commit history for Flame.Cpp, then you'll see that my commit frequency went down quite a bit after the spring and summer of 2015. That's no coincidence; I just didn't need it anymore after that. 

By now, bitrot has probably started creeping into Flame.Cpp. Every time I changed something that was related to Flame's `ICodeGenerator` interface, I patched the C++ back-end to make it compile and run without errors, but I never had the time to properly update it in order to handle the all features I introduced in an _intelligent_ manner. 

So, yes, the C++ back-end needs a makeover, that's for sure. Ideally, it'd also get an updated interface to the C++ standard library, a mini-version of the .NET standard library (just the essentials), and some automated tests. But all of that takes time, and I can only do so many things at a given time. Most of my open source work has been centered around `ecsc` lately.

**Comment by qwertie** [comment]
So here's an obvious question. If Flame is for "managed languages" which normally have garbage collection, how does it target C++ and WebAssembly?

**Comment by jonathanvdc** [comment]
When I say that Flame is for managed languages, what I mean is that it considers reference types, as well as some object-oriented features such as virtual calls, as first-class citizens. A `NewObjectExpression` is treated no differently by the core libraries than a traditionally lower-level construct such as `SizeOfExpression`. The back-end is responsible for translating requests from the IR nodes, such as "create a new object," into code that does the right thing. But the back-end also has total autonomy on _how_ it accomplishes that task. So, to answer your question: 
- The C++ back-end replaces reference type values by `std::shared_ptr<T>` values. I know that's not technically the same as garbage collection, but reference counted pointers are idiomatic C++, so that's why I went with those.
- Right now, the WebAssembly back-end just doesn't support reference types, though I plan on writing a garbage collector to remedy that.

**Comment by qwertie** [comment]
Whew, I'm slow! I'm still finishing up trivia preservation for LES (making sure it round-trips, even just in the common cases, is challenging - the parser is easy as it's one universal algorithm separate from the parser, though it turns out I need to modify it a bit, but the printer is another matter. LESv3 is extra tricky because newline trivia are only allowed in certain places.

**Comment by qwertie** [comment]
Okay, almost done... spent a lot of time reading and writing about stuff related to that crazy U.S. election. Jon, I'll respond to a couple of your posts elsewhere once I commit the trivia-preservation stuff for LESv2&3.

**Comment by qwertie** [comment]
I pushed new code (v1.9.5) last night. Comments & newlines preserved in all parsers and printers now. I plan to make some minor breaking changes and bump the version to 2.0.0 soon. Have you noticed any APIs that you think should work differently?

**Comment by qwertie** [comment]
In [LESv3](http://loyc.net/2016/lesv3-and-wasm.html), I'm wondering if I should change the dot into `#` (int the Loyc tree) so that `.return x + y` means `#return(`'+`(x, y))` - what do you think of that?
- Pro: it's just easier for us to keep the tradition of using `#` for builtins
- Pro: it would allow one to write `#return` to treat the keyword as an identifier instead of having to write ``.return`` with backticks. 
- Con: LES has a delightful visual correspondence between the text syntax and the Loyc tree, and we would be losing that here.

**Comment by qwertie** [comment]
Also, 
- For simplicity's sake I'd like to eliminate block-call expressions like `foo(x) {...}`, because it requires special code to turn off block-call expression support inside a keyword-expression so that when you write `.bar foo(x) {...}`, the braced block is associated with `.bar`, not `foo`. 
- I'm considering changing how continuators work so that parens or braces are not required after a continuator. Notably one could write `else .if ...` instead of having to use a compound continuator "`elseif`".
- I'm considering whether to add a set of keywords like `if`, `while`, `for` (which would be equivalent to `.if` `.while` `.for`)... Disadvantages are the added complexity, and the potential that they would be set in stone for all time.

Thoughts?

**Comment by jonathanvdc** [comment]
Personally, I'm leaning toward rewriting `.return` as `#return` in the LESv3 parser. Admittedly, that's at least in part because I have written code that relies on `#` being the canonical prefix for special identifiers, and I do think that existing LNodes should be printed nicely by LESv3. 

I don't have a strong opinion on block-call expressions and continuators, but I'm fine with any change that simplifies LES. I'm no big fan of keywords, though. I feel like they're constructs that are specific to the programming language that uses them; they don't belong in a universal textual format for code. 

Moreover, there's also the matter of intuitiveness. I typically assume that every keyword has some well-defined meaning, but &ndash; like any other node &ndash; `if`, `while` and `for` don't have specific semantics in LES, which violates my assumption of what keywords do. For example, `while` doesn't make sense in a Haskell syntax tree. If LESv3 is used to parse such a syntax tree, then I feel like it just doesn't make much sense to assign a keyword to something that has no meaning whatsoever. Furthermore, it'll confuse users who expect `if`, `while` and `for` to be legal identifiers.   

All of this comes from my (pretty narrow) use-case, though. I mostly use LES as a handy, human-readable textual representation for Loyc trees, but I don't really _write_ LES myself. So, for my purposes, LES should be as _as simple as possible_ and also _easy to read._ But that's just one use-case. You are much more knowledgeable about LES than I am, and it's up to you to decide in which direction you want LES to evolve.

**Comment by qwertie** [comment]
Here's something fun. In LeMP you can now create macros "on the fly" and I added a new kind of "replace" macro that generates a macro to do the replacement. For example, you know how my version of "static if" has almost no capabilities? I can extend it to support `||` and `&&` by creating four macros like this:

~~~csharp
// It gets `operator||` syntax "for free" because EC# defines operator|| as an alias for @`'||`.
[Passive, ProcessChildrenBefore] replace operator||(false, $x) => $x;
[Passive, ProcessChildrenBefore] replace operator||(true, $x) => true;
[Passive, ProcessChildrenBefore] replace operator&&(false, $x) => false;
[Passive, ProcessChildrenBefore] replace operator&&(true, $x) => $x;

static if ((X `tree==` Y || X `tree==` X) && true)
	then();
else
	otherwise();
~~~

This should make LeMP useful enough to replace some rusty old T4 templates in Loyc.Essentials.dll. It's version 1.9.6 in my personal repo.

**Comment by qwertie** [comment]
So, you don't know of any APIs that you would like to work differently in v2.0.0?

**Comment by jonathanvdc** [comment]
Wow! That's actually really cool. I like how succinct it is. Sort of reminds me of Haskell rewrite rules, but the syntax is way more readable.

> So, you don't know of any APIs that you would like to work differently in v2.0.0?

Okay, I thought about this for a while. Generally, I'm actually pretty happy about the API surface of Loyc that I'm using right now. There is one thing that I think might be worth reviewing, though. And that's [`IMessageSink.Write`](https://github.com/loycnet/ecsharp/blob/358994ced6c7ce3c52716f8bcb5ac1fbdedbe7bb/Core/Loyc.Essentials/Utilities/IMessageSink.cs#L78).

My frustration is that the `context` parameter is an `object`, even though it is always used for the same _concept_ &ndash; in my experience, `context` is _always_ a location in the source code. But the type that is used to represent such a location varies, which is particularly troublesome because not every `IMessageLog` implementation will print directly to the console. These atypical implementations will typically want to convert the `context` to a well-defined representation of a location in the source code. 

If you want an example of where this hurts in practice, you're welcome to take a look at [this paranoid `context`-parsing code](https://github.com/jonathanvdc/ecsc/blob/master/src/Flame.Ecs/CompilerLogMessageSink.cs#L76) in `ecsc`'s `CompilerLogMessageSink`, which translates Loyc diagnostics to Flame diagnostics. And the worst part is that even paranoid code like that is still extremely brittle, because the `context` can be _anything,_ and the actual type of the `context` my message log receives is subject to change without notice. 

Right now, the `context` values I receive are either `SourceRange` or `SourcePos` objects. For my purposes, `SourceRange` has proved itself to be a much more useful data structure than `SourcePos`; the latter are less powerful and are a lot harder to handle if you want to ascertain the _offset_ of the character that the `SourcePos` refers to. Incidentally, that is exactly what `CompilerLogMessageSink.GetContextMarkupNode(SourcePos)` does. 

If `context` were a `SourceRange`, then `Write` would be significantly easier to implement for me.

**Comment by qwertie** [comment]
Well, I'm not going to change the type of `context`, because `IMessageSink` is not designed specifically for compilers; it's also meant as a general-purpose logging system.

However, I will review all uses of `SourcePos` and try to convert them over to `SourceRange`, even if the location is a "point" rather than a "range".

You can use `context = MessageSink.LocationOf(context);` to traverse through `IHasLocation`, and this will include `LNode` since I will change `LNode.Location` to return `Range` instead of `Range.Start`.

**Comment by jonathanvdc** [comment]
> IMessageSink is not designed specifically for compilers; it's also meant as a general-purpose logging system.

Fair enough. But have you considered using a contravariant generic parameter like so?

```cs
public interface IMessageSink<in T>
{
    void Write(string message, T context);
    // ...
}
```

A generic `IMessageSink` gives you a tool to enforce the rule that lexers and parsers must produce a `SourceRange`, and allows me to implement `IMessageSink<SourceRange>` instead of `IMessageSink`. Existing `IMessageSink` implementations can easily be modified to implement `IMessageSink<object>`.

**Comment by qwertie** [comment]
Ahh, good idea... I think I'll change it to `IMessageSink<in TSeverity, in TContext>` and define

~~~csharp
public interface IMessageSink<TContext> : IMessageSink<Severity, TContext>
{
}
public interface IMessageSink : IMessageSink<object>
{
}
~~~

Which is not even a breaking change. Then things like `MessageSink.Console` can be converted to `IMessageSink<SourcePos>`, though sadly `IMessageSink<SourceRange>` won't work since `SourceRange` is a struct.

**Comment by jonathanvdc** [comment]
> Which is not even a breaking change.

Oh, that's a neat trick you pulled there with the generic parameter overloading. 

> sadly `IMessageSink<SourceRange>` won't work since `SourceRange` is a struct.

Right. I didn't think of that until after I posted my comment. Perhaps we could use some kind of reference type wrapper for `SourceRange` instead, then? That is a little convoluted, but I reckon it's neutral performance-wise (`SourceRange` is already being boxed/unboxed ATM), and I'd _really_ like to have a strongly-typed message sink.

**Comment by jonathanvdc** [comment]
In other news, `ecsc` _local_ variables are now identified by `Symbol` keys instead of `string` keys. Consequently, `ecsc` now supports `SymbolPool`-based macro hygiene. 

The `ecsc` standard macros have been updated, and I've taken the liberty of deleting `#builtin_stash_locals` and `#builtin_restore_locals`, as the new hygiene mechanism is much more elegant. As far as I can tell, everything works perfectly!

**Comment by qwertie** [comment]
> Perhaps we could use some kind of reference type wrapper for SourceRange instead, then? That is a little convoluted, but I reckon it's neutral performance-wise (SourceRange is already being boxed/unboxed ATM), and I'd really like to have a strongly-typed message sink.

This reminds me of the fact that I think the CLR's design of `Nullable<T>` was a mistake. Earlier I noticed you can't write a generic class or method that handles both kinds of null (the `Nullable` null and the reference `null`). They should have defined `T?` as a boxed `T`. Now I realize, this design not only would have fit better into generic types, it would have also enabled co/contravariance between struct types and `object`. Sigh.

Loyc.Essentials does have a type to let you box value types without losing access to their members: `Holder<T>`.

There's a problem. There are still some places in my libraries that cannot use `SourceRange` nor `LNode` because they're not reporting about source code. A couple of examples:

- Errors in command-line arguments to LeMP or LLLPG (context: the command-line argument itself)
- The Verbose message "Adding macros in type '{0}'"

Also, in the interest of other clients using LLLPG who are not really interested in Loyc as a whole, I don't want to require a particular kind of context in `BaseLexer` or `BaseParserForList`. [EDIT: also, one of the use cases I designed for is "lexing without allocating memory", which implies not allocating an `ISourceFile` object.]

**Comment by qwertie** [comment]
> ecsc now supports SymbolPool-based macro hygiene.

Cool! Btw I'm having an [interesting chat](https://github.com/dotnet/roslyn/issues/11324) now with someone on the Roslyn team. Since my goal is to improve the world, I need to look for ways of maximizing my impact. So I'm now trying to figure out some kind of macro/extensibility mechanism that the Roslyn/MS people would consider putting in C#. I haven't figured anything out yet... performance and reversibility are key problems.

**Comment by qwertie** [comment]
Hmm, I wonder if I actually shouldn't define `IMessageSink<in TSeverity, in TContext>` because TSeverity is normally `enum Severity`. The standard message sinks can't accept anything else, and if I change them to implement `IMessageSink<object, object>` then a cast to `IMessageSink<Severity, object>` won't work. Plus I don't want to require `Severity` to be boxed in any case, and making (for example) `LogMessage` implement both `IMessageSink<object, object>` and `IMessageSink<Severity, object>` would kinda require me to box the `Severity`, realistically ... plus, some message sinks expect to be able to use operator `<` on two `Severity` values.

**Comment by qwertie** [comment]
Thanks for offering that 'different perspective' btw. Cyrus is a bit merciless...

**Comment by jonathanvdc** [comment]
Oh, well. I guess the Roslyn team are just doing their job in treating your issue as more of a feature request. And anyway, Cyrus took the time to respond to your issue even after you closed it, which I reckon was pretty nice of him. 

He also raises some really good points; refactors should always just work, with or without macros. And that's a pretty fundamental issue. Though I disagree that source generators can be used as a substitute for macros. IMO, this example from [generators.md](https://github.com/dotnet/roslyn/blob/master/docs/features/generators.md) actually looks pretty awful.

```cs
// original.cs:
    partial class C
    {
        void F() { }
        int P { get; set; }
        object this[int index] { get { return null; } }
        event EventHandler E;
    }

// replace.cs:    
    partial class C
    {
        replace void F() { original(); }
        replace int P
        {
            get { return original; }
            set { original += value; } // P.get and P.set
        }
        replace object this[int index]
        {
            get { return original[index]; }
        }
        replace event EventHandler E
        {
            add { original += value; }
            remove { original -= value; }
        }
    }
```

**Comment by qwertie** [comment]
Well, SourceGenerators would seem to have the same problem with refactoring. A source generator can create members that can then called by the original source code - which begs the question, what happens when you try to rename one of them? Perhaps it is an advantage that it's "more obvious" that the rename will fail. SourceGenerators do, admittedly, allow the extract method refactor to work the same as before.

**Comment by qwertie** [comment]
I just released version 2.3.1 - which creates a new assembly Loyc.Math mostly by carving out parts of Loyc.Essentials. Pushed to loycnet/ecsharp too. Probably Flame won't need Loyc.Math but I would be curious to hear if it turns out that you do need a reference to it.

**Comment by qwertie** [comment]
So I was thinking about Semantic Versioning, and how I don't like it. Because switching to it suddenly makes the major version shoot upward very quickly. Especially since, while the Loyc APIs are fairly stable on the whole, there are a LOT of them and I regularly make small backward-incompatible changes. So instead I'm starting to do "Place-shifted semantic versioning", where

- The first number rises very slowly and represents the general "generation" of the product (and when this number is increased, it's a good time to do "spring cleaning" and do more backward-incompatible changes and deprecations than usual.)
- The second number is increased when there are backward-incompatible changes.
- The third number is increased for new APIs and new features, or for small changes that could theoretically be backward-incompatible but probably won't be, like changing the name of a parameter of a public method.
- The fourth number is increased for bug fixes that are (probably) backward-compatible, and refactorings that are (hopefully) invisible to the user.

The corresponding "semantic version" could be defined as `first_num * 10 + second_num).third_num.fourth_num`.

Also, I think it's time to decouple LLLPG versioning from the rest of the repo.

**Comment by jonathanvdc** [comment]
That seems like a perfectly reasonable way to version things. IIRC, LLVM uses a similar versioning scheme. 

I recently removed support for getting the address of the `this` pointer in Flame. IL supports that sort of thing, but it's a pretty useless operation, and it was never used by any front-end. I'm not going to update my major version number now because I got rid of a class without a clear use-case. Sometimes you just have to break things to eliminate technical debt, I guess.

**Comment by jonathanvdc** [comment]
To the best of my knowledge, Flame is not currently using any of Loyc's math-related extensions. Compiler math is usually pretty basic, so there's little need for Loyc.Math at the moment. 

That being said, half-precision and/or quadruple-precision math could be useful if Flame ever targets a back-end that supports those. If that ever happens, Loyc.Math could come in handy.

**Comment by qwertie** [comment]
Loyc.Math doesn't have half-precision or quad-precision floating point math.

I just pushed version 1.4.0.1 to qwertie/ecsharp. I have changed the internal syntax of generic expressions like `X.Y<Z>` to match Roslyn in LES and EC#. Previously the structure was `(X.Y)<Z>`, now it is `X.(Y<Z>)`. I also added some handy new macros like `static matchCode` and `static deconstruct`.

**Comment by jonathanvdc** [comment]
Oh, okay. I thought `FP16` was short for _floating-point, 16 bits,_ but I now realize that it actually meant _fixed-point, 16 decimal bits._ So that sort of had me confused when I wrote my comment earlier. Sorry about that.

**Comment by jonathanvdc** [comment]
I'm sort of on the fence about switching from `(X.Y)<Z>` to `X.(Y<Z>)`. On the plus side, it might just simplify `ecsc`'s logic by a tiny bit. But on the other hand, I feel like `#of(X.Y, Z)` has clearer semantics than `X.#of(Y, Z)`, even if the latter example corresponds to how names are looked up. 

What motivated you to switch representations other than matching Roslyn's AST with the EC# AST? Did you have a use-case or a long-term goal in mind?

**Comment by qwertie** [comment]
I think earlier I had thought about the matter in terms of sequences like `A.B.C<T>` where `A.B.C` is a fully qualified name. In such a case, `T` is logically the parameter of the _whole_ qualified name `A.B.C`.

But when I saw the way Roslyn stores it, I then observed that in an expression like `A.B(1, 2, 3).C<T>`, that's not a qualified name, `C` has nothing to do with `A`, and it doesn't ... feel right, I guess, to parameterize the whole expression `A.B(1,2,3).C` with `T`. So I decided Roslyn's interpretation was more logical in general.

If EC# were the only issue, I wouldn't have done the surprisingly large amount of work it took to change the representation in EC#. This is really about LES - a standard that I want to be used far and wide - and I couldn't really change LES without also changing EC# because I have a bunch of LES code that converts to EC# automatically. So I decided to change both at once.

**Comment by jonathanvdc** [comment]
Fair enough. I get that. Thanks for sharing your reasoning.

On a different note, are you planning to publish NuGet packages for `Loyc.Ecs` and LeMP? Upgrading `ecsc`'s (Loyc) dependencies is a bit of a pain right now because I have to build `Loyc.Ecs` and LeMP from source. That makes it easy for their version numbers to go out of sync with the LoycCore package. 

If you are planning to create NuGet packages for `Loyc.Ecs` and LeMP, then you might be interested in [my PR](https://github.com/qwertie/ecsharp/pull/45).

**Comment by qwertie** [comment]
Just pushed 2.4.2 and published a release. Tomorrow I can start fiddling with NuGet.

**Comment by qwertie** [comment]
Writing articles about Localize and IMessageSink made me decide to make some minor backward-incompatible changes so I'm increasing the version to 2.5.0 (see commit msgs on my PERSONAL repo for details.) Do you have any ideas for backward-incompatible tweaks I might do right now?

**Comment by jonathanvdc** [comment]
Not really. I'm pretty content with the API as-is, and I'm delighted to see that you've introduced `IMessageSink<TContext>`.

**EDIT:** fixed a typo

**Comment by qwertie** [comment]
> This version of Flame/dsc adds support for user-defined conversions and fixes some dsc-specific operator overloading bugs with addition, left shifts, and right shifts.

I was a bit surprised to see this, since I just noticed a few days ago that EC# cannot parse "operator<<" and "operator>>" (the reason had to do with `<<` and `<<` being two tokens, unlike all other operators) (I fixed it soon afterward). Didn't you encounter that bug too?

**Comment by jonathanvdc** [comment]
Heh. That's a funny coincidence. The bug I fixed was that the D# front-end refused to consider user-defined operators when it saw `a >> b` or `a << b` &ndash; so it's more of a call-site bug than a declaration-site problem.

On the EC# parser bug you mentioned: I just reproduced it in my `ecsc` test suite and I'll commit the updated test once the next EC# parser NuGet package is released. I hadn't stumbled into it until you brought it up, though.

In other news: I wrote a [tiny tutorial](https://jonathanvdc.gitbooks.io/flame-brainfuck/content) on how to build a Brainfuck compiler with Flame. It's mostly just an introduction to using Flame, so most of it is explaining basic concepts. It has a couple of defects: it might not be as gentle as I'd hoped it would be, and it's mostly a discussion of _how_ to build a compiler with Flame, rather than an explanation of _why_ someone might want to use Flame in the first place. I guess it needs some more work. Regardless, I think it's a step up from the previous situation, i.e., no tutorials at all. So there's that.
